_posts/all-new-goodies-in-servicecomposer.md

```diff
@@ -0,0 +1,179 @@
+
+## Versions 4.0 and 4.1
+
+Versions [4.0](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/releases/tag/4.0.0) and [4.1](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/releases/tag/4.1.0) continue the good work users nudged me to do.
+
+### System.Text.Json replaced Newtonsoft.Json
+
+Version 4.0 started with a breaking change, which incidentally was also a requested "feature." System.Text.Json replaced Newtonsoft.Json for all the internal serialization needs. Tests prove that the externally visible behavior did not change. Still, it's a breaking change, and thus, it deserves a major release.
+
+### Declarative model binding
+
+ServiceComposer added support for [model binding](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/blob/master/docs/model-binding.md) in version 1.9, a long time ago in March 2021. Since then, little to nothing changed. The essence of model binding is that it lets users define models for incoming requests, for example, a class like the following:
+
+```csharp
+class RequestModel
+{
+    [FromRoute(Name = "id")] public int Id { get; set; }
+    [FromBody] public BodyModel Body { get; set; }
+}
+```
+
+And then, in the composition handler, bind the above model:
+
+```csharp
+[HttpPost("/sample/{id}")]
+public async Task Handle(HttpRequest request)
+{
+    var requestModel = await request.Bind<RequestModel>();
+    var body = requestModel.Body;
+    var aString = body.AString;
+    var id = requestModel.Id;
+
+    //use values as needed
+}
+```
+
+The above code is nothing wrong, but unfortunately, it doesn't play nicely with endpoint filters. The endpoint filters API exposes an arguments list representing the binding results performed by ASP.Net before invoking the HTTP handling pipeline. The need to populate arguments means binding must happen earlier in the process and cannot be performed in the composition handler if users also want to use endpoint filters.
+
+With [declarative model binding](https://github.com/ServiceComposer/ServiceComposer.AspNetCore/blob/master/docs/model-binding.md#declarative-model-binding), the above composition handler looks like the following:
+
+```csharp
+[HttpPost("/sample/{id}")]
+[BindFromBody<BodyModel>]
+[BindFromRoute<int>(routeValueKey: "id")]
+public Task Handle(HttpRequest request)
+{
+    var arguments = request.GetCompositionContext().GetArguments(this);
+    var body = arguments.Argument<BodyModel>();
+    var id = arguments.Argument<int>(name: "id");
+
+    return Task.CompletedTask;
+}
+```
+
+The `RequestModel` wrapper class is no longer needed. The code declares the models for binding, using attributes, and later, it can access them via the arguments collection exposed by the composition context. There are attributes for all the supported binding sources, such as the query string or request headers. 
```

Comment:
```suggestion
The `RequestModel` wrapper class is no longer needed. The code declares the models for binding using attributes and later, it can access them via the arguments collection exposed by the composition context. There are attributes for all the supported binding sources, such as the query string or request headers. 
```